csharp/ql/src/semmle/code/csharp/exprs/Call.qll

```diff
@@ -555,15 +566,159 @@ class DelegateCall extends Call, @delegate_invocation_expr {
+ * `set_Prop()` is not the access `Prop`, but rather the assignment `Prop = i + 1`.
+ * (Using `Prop` as the call to `set_Prop()` will yield an incorrect control flow
+ * graph, where `set_Prop()` is called before `i + 1` is evaluated.)
+ */
+private module AccessorCallImpl {
+  /**
+   * Holds if `e` is a tuple assignment with multiple setter assignments,
+   * for example `(Prop1, Prop2) = (0, 1)`, where both `Prop1` and `Prop2`
+   * are properties.
+   *
+   * In such cases, we cannot associate the compound assignment with the
+   * setter calls, so we instead revert to representing the calls via the
+   * individual accesses on the left hand side.
+   */
+  private predicate multiTupleSetter(Expr e) {
+    strictcount(AssignableDefinitions::TupleAssignmentDefinition def |
+      def.getExpr() = e and
+      def.getTargetAccess().getTarget() instanceof DeclarationWithGetSetAccessors
+    ) > 1
+  }
+
+  abstract class AccessorCallImpl extends Expr {
+    abstract Access getAccess();
+
+    abstract Accessor getTarget();
+
+    abstract Expr getArgument(int i);
+  }
+
+  abstract class PropertyCallImpl extends AccessorCallImpl {
+    PropertyAccess pa;
+
+    override PropertyAccess getAccess() { result = pa }
+
+    abstract override Accessor getTarget();
+
+    abstract override Expr getArgument(int i);
+  }
+
+  class PropertyGetterCall extends PropertyCallImpl, @property_access_expr {
+    PropertyGetterCall() {
+      this instanceof AssignableRead and
+      pa = this
+    }
+
+    override Getter getTarget() { result = pa.getProperty().getGetter() }
+
+    override Expr getArgument(int i) { none() }
+  }
+
+  class PropertySetterCall extends PropertyCallImpl {
+    PropertySetterCall() {
+      exists(AssignableDefinition def | pa = def.getTargetAccess() |
+        (if multiTupleSetter(def.getExpr()) then this = pa else this = def.getExpr()) and
+        not def.getExpr().(AssignOperation).hasExpandedAssignment()
+      )
+    }
+
+    override Expr getArgument(int i) {
+      i = 0 and
+      result = AssignableInternal::getAccessorCallValueArgument(pa)
+    }
+
+    override Setter getTarget() { result = pa.getProperty().getSetter() }
+  }
+
+  abstract class IndexerCallImpl extends AccessorCallImpl {
+    IndexerAccess ia;
+
+    override IndexerAccess getAccess() { result = ia }
+
+    abstract override Accessor getTarget();
+
+    abstract override Expr getArgument(int i);
+  }
+
+  class IndexerGetterCall extends IndexerCallImpl, @indexer_access_expr {
+    IndexerGetterCall() {
+      this instanceof AssignableRead and
+      ia = this
+    }
+
+    override Getter getTarget() { result = ia.getIndexer().getGetter() }
+
+    override Expr getArgument(int i) { result = ia.getIndex(i) }
+  }
+
+  class IndexerSetterCall extends IndexerCallImpl {
+    IndexerSetterCall() {
+      exists(AssignableDefinition def | ia = def.getTargetAccess() |
+        (if multiTupleSetter(def.getExpr()) then this = ia else this = def.getExpr()) and
```

Comment:
Just to clarify - something like this could be possible?
```
(a[0], a[1]) = (4,5);
```

Reply:
Might be worth adding a test for that (if not done already).

Reply:
Luckily there is a test case already (this is how I discovered the problem in the first case): https://github.com/Semmle/ql/blob/master/csharp/ql/test/library-tests/arguments/arguments.cs#L56.